packages/wallet/src/browser/index.ts

```diff
@@ -19,147 +14,123 @@ import {
-  );
+  async getUsedAddresses() {
+    const usedAddresses = await this.extension.getUsedAddresses();
+    return usedAddresses.map(deserializeAddress);
+  }
 
-const fetchCIP30Network: (
-  extensionCIP30Instance: BroswerExtensionCIP30Api
-) => T.Task<CIP30Network> = (extensionCIP30Instance) =>
-  pipe(
-    T.Do,
-    T.bind("networkId", () =>
-      pipe(() => extensionCIP30Instance.getNetworkId())
-    ),
-    T.map(({ networkId }) => (networkId == 1 ? "Mainnet" : "Testnets"))
-  );
+  async getCollaterals() {
+    const collaterals =
+      (await this.extension.experimental.getCollateral()) ?? [];
+    return collaterals.map(deserializeTxOutRef);
+  }
 
-const fetchUTxOs: (
-  extensionCIP30Instance: BroswerExtensionCIP30Api
-) => T.Task<TxOutRef[]> = (extensionCIP30Instance) =>
-  pipe(
-    T.Do,
-    T.bind("uxtxos", () => pipe(() => extensionCIP30Instance.getUtxos())),
-    T.map(({ uxtxos }) =>
-      uxtxos == undefined ? [] : pipe(uxtxos, A.map(deserializeCollateral))
-    )
-  );
-const fetchCollaterals: (
-  extensionCIP30Instance: BroswerExtensionCIP30Api
-) => T.Task<TxOutRef[]> = (extensionCIP30Instance) =>
-  pipe(
-    T.Do,
-    T.bind("collaterals", () =>
-      pipe(() => extensionCIP30Instance.experimental.getCollateral())
-    ),
-    T.map(({ collaterals }) =>
-      collaterals == undefined
-        ? []
-        : pipe(collaterals, A.map(deserializeCollateral))
-    )
-  );
+  async getUTxOs() {
+    const utxos = (await this.extension.getUtxos()) ?? [];
+    return utxos.map(deserializeTxOutRef);
+  }
 
-const deserializeAddress: (addressHex: string) => AddressBech32 = (
-  addressHex
-) =>
-  pipe(
-    C.Address.from_bytes(hex.decode(addressHex)).to_bech32(undefined),
-    addressBech32
-  );
+  async getCIP30Network(): Promise<CIP30Network> {
+    const networkId = await this.extension.getNetworkId();
+    return networkId == 1 ? "Mainnet" : "Testnets";
+  }
+
+  async getTokens() {
+    const balances = await this.extension.getBalance();
+    return valueToTokens(deserializeValue(balances));
+  }
+
+  async getLovelaces(): Promise<bigint> {
+    const balances = await this.extension.getBalance();
+    return valueToLovelaces(deserializeValue(balances));
+  }
+}
+
+/**
+ * Returns an instance of the browser wallet API for the specified wallet.
+ * @param walletName - The name of the wallet to get an instance of.
+ * @returns An instance of the BrowserWalletAPI class.
+ */
+export async function createBrowserWallet(
+  walletName: SupportedWallet
+): Promise<WalletAPI> {
+  if (getAvailableWallets().includes(walletName)) {
+    const extension = await window.cardano[walletName.toLowerCase()].enable();
+    return new BrowserWalletAPI(extension);
+  } else {
+    throw new Error(`Wallet ${walletName} is not available in the browser`);
+  }
+}
+
+/**
+ * Get a list of the available wallets installed in the browser
+ */
+export function getAvailableWallets(): SupportedWallet[] {
```

Comment:
I wasn't thinking implementing this function that way... Like hard core filtering them like that...It is too rigid imo